test/integration/auth.test.js

```diff
@@ -254,4 +256,296 @@ describe("auth", function () {
+    ];
+    const fetchStub = sinon.stub(global, "fetch").resolves(new Response(JSON.stringify(fakeEmails)));
+
+    sinon.stub(passport, "authenticate").callsFake((strategy, options, callback) => {
+      callback(null, "accessToken", {
+        username: "github-user",
+        displayName: "GitHub User",
+        _json: { email: null, created_at: "2022-01-01" },
+        id: 12345,
+      });
+      return (req, res, next) => {};
+    });
+
+    const res = await chai
+      .request(app)
+      .get(`/auth/github/callback`)
+      .query({ code: "codeReturnedByGithub", state: rdsUrl })
+      .redirects(0);
+    expect(res).to.have.status(302);
+
+    const fetchArgs = fetchStub.getCall(0).args;
+    expect(fetchArgs[0]).to.equal("https://api.github.com/user/emails");
+    expect(fetchArgs[1].headers.Authorization).to.equal("token accessToken");
+  });
+
+  it("should return google call back URL", async function () {
+    const googleOauthURL = generateGoogleAuthRedirectUrl({});
+    const res = await chai.request(app).get("/auth/google/login?dev=true").redirects(0);
+    expect(res).to.have.status(302);
+    expect(res.headers.location).to.equal(googleOauthURL);
+  });
+
+  it("should return google call back URL with redirectUrl", async function () {
+    const RDS_MEMBERS_SITE_URL = "https://members.realdevsquad.com";
+    const googleOauthURL = generateGoogleAuthRedirectUrl({ state: RDS_MEMBERS_SITE_URL });
+    const res = await chai
+      .request(app)
+      .get("/auth/google/login?dev=true")
+      .query({ redirectURL: RDS_MEMBERS_SITE_URL })
+      .redirects(0);
+    expect(res).to.have.status(302);
+    expect(res.headers.location).to.equal(googleOauthURL);
+  });
+
+  it("should redirect the google user to new sign up flow if they are have incomplete user details true", async function () {
+    const redirectURL = "https://my.realdevsquad.com/new-signup";
+
+    sinon.stub(passport, "authenticate").callsFake((strategy, options, callback) => {
+      callback(null, "accessToken", googleUserInfo[0]);
+      return (req, res, next) => {};
+    });
+
+    const res = await chai
+      .request(app)
+      .get("/auth/google/callback")
+      .query({ code: "codeReturnedByGoogle" })
+      .redirects(0);
+    expect(res).to.have.status(302);
+    expect(res.headers.location).to.equal(redirectURL);
+  });
+
+  it("should redirect the google user to the goto page on successful login, if user has incomplete user details false", async function () {
+    await addUserToDBForTest(googleUserInfo[1]);
+    const rdsUiUrl = new URL(config.get("services.rdsUi.baseUrl")).href;
+    sinon.stub(passport, "authenticate").callsFake((strategy, options, callback) => {
+      callback(null, "accessToken", googleUserInfo[0]);
+      return (req, res, next) => {};
+    });
+
+    const res = await chai
+      .request(app)
+      .get("/auth/google/callback")
+      .query({ code: "codeReturnedByGoogle", state: rdsUiUrl })
```

Comment:
What if state is not passed? and code is not passed

Reply:
If the state is not passed, the user will be redirected to the RDS main site. If the code is missing or invalid, the request will return a 401 Unauthorized error because the code is necessary to exchange for an access token and complete the OAuth flow.

I’ve also written tests to verify both of these scenarios.